### PR TITLE
Split ITransactionalSession and IBatchSession APIs

### PR DESCRIPTION
**Pull request by timbussmann**
Removes the sync storage support from the non-outbox connected session so that it only contains the batch message operations functionality.

### CODE REVIEW

File: src/NServiceBus.TransactionalSession/TransactionalSessionFeature.cs [hunk 1]
```diff
@@ -28,30 +28,28 @@ protected override void Setup(FeatureConfigurationContext context)
...
             var sessionCaptureTask = new SessionCaptureTask();
             context.RegisterStartupTask(sessionCaptureTask);
-            context.Services.AddScoped<ITransactionalSession>(sp =>
-            {
-                var physicalLocalQueueAddress = sp.GetRequiredService<ITransportAddressResolver>()
-                    .ToTransportAddress(localQueueAddress);
 
-                if (isOutboxEnabled)
+
+            context.Services.AddScoped<IBatchSession>(sp => new TransactionalSession(
```
Comment (timbussmann) [inline src/NServiceBus.TransactionalSession/TransactionalSessionFeature.cs:33]: let's just register than when the outbox is disabled.

File: src/NServiceBus.TransactionalSession/ITransactionalSession.cs [hunk 2]
```diff
@@ -1,72 +1,16 @@
...
-        /// Instantiates a message of type T and publishes it.
-        /// </summary>
-        /// <typeparam name="T">The type of message, usually an interface.</typeparam>
-        /// <param name="messageConstructor">An action which initializes properties of the message.</param>
-        /// <param name="publishOptions">Specific options for this event.</param>
-        /// <param name="cancellationToken">A <see cref="CancellationToken"/> to observe.</param>
-        Task Publish<T>(Action<T> messageConstructor, PublishOptions publishOptions, CancellationToken cancellationToken = default);
-
         /// <summary>
         /// Gets the synchronized storage session for processing the current message. NServiceBus makes sure the changes made
```
Comment (tmasternak) [inline src/NServiceBus.TransactionalSession/ITransactionalSession.cs:11]:
```suggestion
        /// Gets the synchronized storage session associated with the <see cref="ITransactionalSession" />. Operations performed on the storage session and messaging operations (send and publish) for a single unit of work and will be atomically committed.  
```

File: src/NServiceBus.TransactionalSession/ITransactionalSession.cs [hunk 3]
```diff
@@ -1,72 +1,16 @@
...
         /// via this session will be persisted before the message receive is acknowledged.
```
Comment (tmasternak) [inline src/NServiceBus.TransactionalSession/ITransactionalSession.cs:12]:
```suggestion
```

File: src/NServiceBus.TransactionalSession/OutboxTransactionalSession.cs [hunk 4]
```diff
@@ -11,19 +11,34 @@
...
         }
 
+        public ISynchronizedStorageSession SynchronizedStorageSession
+        {
+            get
+            {
+                if (!IsOpen)
+                {
+                    throw new InvalidOperationException(
+                        "Before accessing the SynchronizedStorageSession, make sure to open the session by calling the `Open`-method.");
```
Comment (tmasternak) [inline src/NServiceBus.TransactionalSession/OutboxTransactionalSession.cs:35]:
```suggestion
                        $"Before accessing the {nameof(SynchronizedStorageSession)}, make sure to open the session by calling the `Open`-method.");
```

File: src/NServiceBus.TransactionalSession/TransactionalSession.cs [hunk 5]
```diff
@@ -2,30 +2,20 @@
 {
     using System.Threading;
     using System.Threading.Tasks;
-    using Persistence;
     using Transport;
 
-    class TransactionalSession : TransactionalSessionBase
+    class TransactionalSession : TransactionalSessionBase, IBatchSession
```
Comment (tmasternak) [inline src/NServiceBus.TransactionalSession/TransactionalSession.cs:7]: Should this be named `BatchSession`?

File: src/NServiceBus.TransactionalSession/TransactionalSessionBase.cs [hunk 6]
```diff
@@ -4,36 +4,19 @@ namespace NServiceBus.TransactionalSession
     using System.Threading;
     using System.Threading.Tasks;
     using Extensibility;
-    using Persistence;
     using Transport;
 
-    abstract class TransactionalSessionBase : ITransactionalSession
+    abstract class TransactionalSessionBase
```
Comment (tmasternak) [inline src/NServiceBus.TransactionalSession/TransactionalSessionBase.cs:9]: Should this be called `SessionBase`?

**Review by tmasternak — DISMISSED** [review]

File: src/NServiceBus.TransactionalSession/ITransactionalSession.cs [hunk 7]
```diff
@@ -11,8 +11,8 @@
     public interface ITransactionalSession : IBatchedMessageSession
     {
         /// <summary>
-        /// Gets the synchronized storage session for processing the current message. NServiceBus makes sure the changes made
-        /// via this session will be persisted before the message receive is acknowledged.
+        /// Gets the synchronized storage session for processing the current message.
+        /// Operations performed on the storage session and message operations (send and publish) for a single unit of work and will be atomically committed.
```
Comment (lailabougria) [inline src/NServiceBus.TransactionalSession/ITransactionalSession.cs:15]:
```suggestion
        /// Gets the synchronized storage session for processing the current message, which is used to atomically execute any operations performed on the storage session, as well as all the message operations (send and publish)
```

File: src/NServiceBus.TransactionalSession/ITransactionalSession.cs [hunk 8]
```diff
@@ -70,10 +22,12 @@ public interface ITransactionalSession : IDisposable
         string SessionId { get; }
 
         /// <summary>
-        /// Commit the session by applying all message and synchronized storage operation in an atomic manner.
+        /// Opens the transaction session.
```
Comment (lailabougria) [inline src/NServiceBus.TransactionalSession/ITransactionalSession.cs:25]:
```suggestion
        /// Opens the transactional session.
```

**Review by lailabougria — APPROVED** [review]
Few small suggestions, otherwise 👍🏼

File: src/NServiceBus.TransactionalSession/ITransactionalSession.cs (hunk 7 above)
Comment (timbussmann) [reply src/NServiceBus.TransactionalSession/ITransactionalSession.cs:15]: >  for processing the current message

I don't understand this part

Comment (timbussmann) [reply src/NServiceBus.TransactionalSession/ITransactionalSession.cs:15]: I think having shorter sentences will be easier to read.

File: src/NServiceBus.TransactionalSession/ITransactionalSession.cs [hunk 9]
```diff
@@ -11,8 +11,8 @@
     public interface ITransactionalSession : IBatchedMessageSession
     {
         /// <summary>
-        /// Gets the synchronized storage session for processing the current message. NServiceBus makes sure the changes made
-        /// via this session will be persisted before the message receive is acknowledged.
+        /// Gets the synchronized storage session for processing the current message.
```
Comment (timbussmann) [inline src/NServiceBus.TransactionalSession/ITransactionalSession.cs:14]:
```suggestion
        /// Gets the synchronized storage session for this transactional session.
```

File: src/NServiceBus.TransactionalSession/ITransactionalSession.cs (hunk 7 above)
Comment (timbussmann) [inline src/NServiceBus.TransactionalSession/ITransactionalSession.cs:15]:
```suggestion
        /// Storage operations are committed atomically with the message operations executed on this session.
```

**Pull request merged**